boefjes/tools/upgrade_v1_16_0.py

```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python3
+
+"""Migration script for v1.16.0 due to a bug in the garbage collection. To be removed in later versions.
+https://github.com/minvws/nl-kat-coordination/issues/2875"""
+
+import json
+import logging.config
+import sys
+from datetime import datetime, timezone
+from pathlib import Path
+
+import click
+from httpx import HTTPStatusError
+from sqlalchemy.orm import sessionmaker
+
+from boefjes.api import get_bytes_client
+from boefjes.config import settings
+from boefjes.dependencies.plugins import PluginService
+from boefjes.job_handler import get_octopoes_api_connector
+from boefjes.local_repository import get_local_repository
+from boefjes.models import Boefje
+from boefjes.sql.config_storage import create_config_storage
+from boefjes.sql.db import get_engine
+from boefjes.sql.organisation_storage import create_organisation_storage
+from boefjes.sql.plugin_storage import create_plugin_storage
+from boefjes.storage.interfaces import OrganisationStorage
+from octopoes.models.origin import OriginType
+
+sys.path.append(str(Path(__file__).resolve().parent.parent))
+
+
+with settings.log_cfg.open() as f:
+    logging.config.dictConfig(json.load(f))
+
+logger = logging.getLogger(__name__)
+
+
+def upgrade(organisation_repository: OrganisationStorage, valid_time: datetime | None = None) -> tuple[int, int]:
+    """
+    Perform the migration for all organisations in the database. The happy flow in this script is idempotent,
+    meaning that it can be rerun until there are no, or only expected, exceptions.
+    """
+    if valid_time is None:
+        valid_time = datetime.now(timezone.utc)
+
+    bytes_client = get_bytes_client()
+    bytes_client.login()
+
+    total_failed = 0
+    total_processed = 0
+
+    organisations = organisation_repository.get_all()
+    logger.info("Processing %s organisations in total", len(organisations))
+
+    for organisation_id in organisations:
+        connector = get_octopoes_api_connector(organisation_id)
+        logger.info("Processing organisation [organization_id=%s]", organisation_id)
+
+        failed, processed = migrate_organisation(bytes_client, connector, organisation_id, valid_time)
+        total_failed += failed
+        total_processed += processed
+
+        logger.info("Processed organisation [total_processed=%s, total_failed=%s]", processed, failed)
+
+    logger.info("Finished migration [total_processed=%s, total_failed=%s]", total_processed, total_failed)
+
+    return total_processed, total_failed
+
+
+def migrate_organisation(bytes_client, connector, organisation_id, valid_time) -> tuple[int, int]:
+    """
+    For each organisation, we paginate through the origin API, find the matching normalizer meta in Bytes,
+    and set the source_method to the boefje id. Then update the origin, i.e. save it and delete the old one.
+    """
+
+    failed = 0
+    processed = 0
+
+    offset = 0
+    page_size = 200
+
+    boefjes_for_normalizer = collect_boefjes_per_normalizer()
```

Comment:
This call isn't organisation specific so it isn't necessary to call this for each organisation.

Reply:
Ah indeed, I think first versions were, while still using `get_all()`